### PR TITLE
[SPARK-39078][SQL] Support UPDATE commands with DEFAULT values

### PR DESCRIPTION
**Pull request by dtenedor**
## What changes were proposed in this pull request?

DEFAULT columns are now supported in UPDATE commands.

Example:

```
CREATE TABLE t(
  i STRING,
  c CHAR(5),
  d VARCHAR(5),
  e STRING DEFAULT 'abc',
  f BIGINT DEFAULT 42)
  USING DELTA;
INSERT INTO t VALUES('1', '12345', '67890', 'def', 31);
UPDATE t SET c = DEFAULT WHERE i = '1';
SELECT * FROM t;
> "1", NULL, "67890", "def", 31
UPDATE t SET e = DEFAULT WHERE i = '1';
SELECT * FROM t;
> "1", NULL, "67890", "abc", 31
UPDATE t SET e = 'ghi' WHERE i = '1';
SELECT * FROM t;
> "1", NULL, "67890", "ghi", 31
```

### Why are the changes needed?

This makes UPDATE commands easier to use.

### Does this PR introduce _any_ user-facing change?

Yes.

## How was this patch tested?

This PR adds new unit test coverage.

### CODE REVIEW

**Comment by dtenedor** [comment]
Hi @gengliangwang when you're ready, this is the OSS PR to support DEFAULT values in UPDATE commands. It's based on https://github.com/apache/spark/pull/36398 and tested with unit tests in `PlanResolutionSuite` per recommendation from @cloud-fan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala [hunk 1]
```diff
@@ -55,11 +55,21 @@ object ResolveDefaultColumns {
...
   val CURRENT_DEFAULT_COLUMN_NAME = "DEFAULT"
-  // Return a more descriptive error message if the user tries to nest the DEFAULT column reference
-  // inside some other expression, such as DEFAULT + 1 (this is not allowed).
-  val DEFAULTS_IN_EXPRESSIONS_ERROR = "Failed to execute INSERT INTO command because the " +
-    "VALUES list contains a DEFAULT column reference as part of another expression; this is " +
-    "not allowed"
+  // Return a more descriptive error message if the user tries to nest a DEFAULT column reference
+  // inside some other expression (such as DEFAULT + 1) in an INSERT INTO command's VALUES list;
+  // this is not allowed.
+  val DEFAULTS_IN_COMPLEX_EXPRESSIONS_IN_INSERT_VALUES = "Failed to execute INSERT INTO command " +
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:61]: Let's move the error messages to QueryCompilationErrors

**Review by dtenedor — COMMENTED** [review]
Thanks for your review!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 2]
```diff
@@ -131,11 +135,41 @@ case class ResolveDefaultColumns(
...
+    }.getOrElse(i)
+  }
+
+  /**
+   * Resolves DEFAULT column references for an UPDATE command.
+   */
+  private def resolveDefaultColumnsForUpdate(u: UpdateTable): LogicalPlan = {
+    // Return a more descriptive error message if the user tries to use a DEFAULT column reference
+    // inside an UPDATE command's WHERE clause; this is not allowed.
+    u.condition.map { c: Expression =>
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:152]: nit: Let's use `u.condition.foreach` instead of `u.condition.map`

Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:152]: nit: Let's use `u.condition.foreach` instead of `u.condition.map`

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:152]: Actually, I can see this from IDE(Ultimate Edition):
image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 3]
```diff
@@ -131,11 +135,41 @@ case class ResolveDefaultColumns(
...
+      if (c.find(isExplicitDefaultColumn).isDefined) {
+        throw QueryCompilationErrors.defaultReferencesNotAllowedInUpdateWhereClause()
+      }
+    }
+    val schema: StructType = getSchemaForTargetTable(u.table).getOrElse(return u)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:157]: nit: let's avoid `return` in the closure.

Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:157]: nit: let's avoid `return` in the closure.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 4]
```diff
@@ -91,30 +94,31 @@ case class ResolveDefaultColumns(
    * [[insertsFromInlineTable]] method.
    */
   private def resolveDefaultColumnsForInsertFromInlineTable(i: InsertIntoStatement): LogicalPlan = {
+    val insertTableSchemaWithoutPartitionColumns: StructType =
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:97]: This is not related to this PR. It would be better if you can open a new refactor PR for it.

Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:97]: This is not related to this PR. It would be better if you can open a new refactor PR for it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 5]
```diff
@@ -91,30 +94,31 @@ case class ResolveDefaultColumns(
...
+      getInsertTableSchemaWithoutPartitionColumns(i)
+        .getOrElse(return i)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:99]: nit: let's avoid return in the closure.

Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:99]: nit: let's avoid return in the closure.

**Comment by gengliangwang** [comment]
LGTM overall

**Comment by gengliangwang** [comment]
LGTM overall

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 6]
```diff
@@ -98,44 +101,74 @@ case class ResolveDefaultColumns(
...
+
+  /**
+   * Resolves DEFAULT column references for an UPDATE command.
+   */
+  private def resolveDefaultColumnsForUpdate(u: UpdateTable): LogicalPlan = {
+    // Return a more descriptive error message if the user tries to use a DEFAULT column reference
+    // inside an UPDATE command's WHERE clause; this is not allowed.
+    u.condition.foreach { c: Expression =>
+      if (c.find(isExplicitDefaultColumn).isDefined) {
+        throw QueryCompilationErrors.defaultReferencesNotAllowedInUpdateWhereClause()
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:151]: Let's have a test case for this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 7]
```diff
@@ -1035,6 +1060,41 @@ class PlanResolutionSuite extends AnalysisTest {
...
+          assert(name.name == "name")
+          assert(age.name == "age")
+
+        case _ => fail("Expect UpdateTable, but got:\n" + parsed5.treeString)
+      }
+
+      parsed6 match {
+        case UpdateTable(
+        AsDataSourceV2Relation(_),
+        Seq(Assignment(i: AttributeReference, AnsiCast(Literal(null, _), IntegerType, _)),
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:1079]: Hmm, probably I missed this in the previous reviews. So we will insert a NULL value for a default expression if the default column value is not defined, right?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 8]
```diff
@@ -1035,6 +1060,41 @@ class PlanResolutionSuite extends AnalysisTest {
 
         case _ => fail("Expect UpdateTable, but got:\n" + parsed4.treeString)
       }
+
+      parsed5 match {
+        case UpdateTable(
+        AsDataSourceV2Relation(_),
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:1066]: nit: indent

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 9]
```diff
@@ -1035,6 +1060,41 @@ class PlanResolutionSuite extends AnalysisTest {
...
+        None) =>
+          assert(name.name == "name")
+          assert(age.name == "age")
+
+        case _ => fail("Expect UpdateTable, but got:\n" + parsed5.treeString)
+      }
+
+      parsed6 match {
+        case UpdateTable(
+        AsDataSourceV2Relation(_),
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:1078]: nit: indent

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala [hunk 10]
```diff
@@ -1035,6 +1060,41 @@ class PlanResolutionSuite extends AnalysisTest {
...
+        None) =>
+          assert(i.name == "i")
+          assert(s.name == "s")
+
+        case _ => fail("Expect UpdateTable, but got:\n" + parsed6.treeString)
+      }
+
+      parsed7 match {
+        case UpdateTable(
+        _,
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:1090]: nit: indent

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala (hunk 6 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:151]: SG, done.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala (hunk 8 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala (hunk 9 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala (hunk 7 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: Yes: when resolving DEFAULT column references, the analyzer will insert literal NULL values if the corresponding table does not define an explicit default value for that column. This is intended.

I left a comment here to this effect in case it's confusing to any future readers.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala (hunk 10 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala]: Done.

**Review by dtenedor — COMMENTED** [review]
Thanks for your review!

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master
